pyro/infer/autoguide/guides.py

```diff
@@ -29,37 +29,47 @@ def model():
 from pyro.infer.enum import config_enumerate
 from pyro.nn import AutoRegressiveNN
 from pyro.ops.hessian import hessian
+from pyro.params import ConstrainedModule, ConstrainedParameter
 from pyro.poutine.util import prune_subsample_sites
 
 
-class AutoGuide(object):
+class AutoGuide(ConstrainedModule):
     """
     Base class for automatic guides.
 
-    Derived classes must implement the :meth:`__call__` method.
+    Derived classes must implement the :meth:`forward` method, with the
+    same ``*args, **kwargs`` as the base ``model``.
 
     Auto guides can be used individually or combined in an
     :class:`AutoGuideList` object.
 
     :param callable model: a pyro model
     :param str prefix: a prefix that will be prefixed to all param internal sites
+        as part of the module name.
     """
 
     def __init__(self, model, prefix="auto"):
+        super().__init__()
         self.master = None
         self.model = model
         self.prefix = prefix
         self.prototype_trace = None
         self._plates = {}
 
-    def __call__(self, *args, **kwargs):
+    def call(self, *args, **kwargs):
         """
-        A guide with the same ``*args, **kwargs`` as the base ``model``.
+        Method that calls :meth:`forward` and returns parameter values of the
+        guide as a `tuple` instead of a `dict`, which is a requirement for
+        JIT tracing. Unlike :meth:`forward`, this method can be traced by
+        :func:`torch.jit.trace_module`.
 
-        :return: A dict mapping sample site name to sampled value.
-        :rtype: dict
+        .. warning::
+            This method may be removed once PyTorch JIT tracer starts accepting
+            `dict` as valid return types. See
+            `issue <https://github.com/pytorch/pytorch/issues/27743>_`.
         """
-        raise NotImplementedError
+        result = self(*args, **kwargs)
+        return tuple(result.values())
```

Comment:
This has undefined order. Instead
```py
return tuple(v for k, v in sorted(result.items()))
```

Reply:
I overlooked this working with `nn.Module` where everything is an OrderedDict. Will fix.